src/shapes/Textbox.ts

```diff
@@ -314,84 +370,56 @@ export class Textbox extends IText {
-      splitByGrapheme = this.splitByGrapheme,
-      graphemeLines = [],
-      words = splitByGrapheme
-        ? this.graphemeSplit(_line)
-        : this.wordSplit(_line),
-      infix = splitByGrapheme ? '' : ' ';
-
-    let lineWidth = 0,
-      line = [],
-      // spaces in different languages?
-      offset = 0,
-      infixWidth = 0,
-      largestWordWidth = 0,
-      lineJustStarted = true;
+  ) {
+    const measurements = this.measureLineForWrapping(line, lineIndex);
+
     // fix a difference between split and graphemeSplit
-    if (words.length === 0) {
-      words.push([]);
+    if (!measurements) {
+      return [[]];
     }
-    desiredWidth -= reservedSpace;
-    // measure words
-    const data = words.map((word) => {
-      // if using splitByGrapheme words are already in graphemes.
-      word = splitByGrapheme ? word : this.graphemeSplit(word);
-      const width = this._measureWord(word, lineIndex, offset);
-      largestWordWidth = Math.max(width, largestWordWidth);
-      offset += word.length + infix.length;
-      return { word, width };
-    });
+
+    const { data, minWidth } = measurements;
+    const additionalSpace = this._getWidthOfCharSpacing();
 
     const maxWidth = Math.max(
-      desiredWidth,
-      largestWordWidth,
+      desiredWidth - reservedSpace,
+      minWidth,
       this.dynamicMinWidth
     );
-    // layout words
-    offset = 0;
-    let i;
-    for (i = 0; i < words.length; i++) {
-      const word = data[i].word;
-      const wordWidth = data[i].width;
-      offset += word.length;
-
-      lineWidth += infixWidth + wordWidth - additionalSpace;
-      if (lineWidth > maxWidth && !lineJustStarted) {
-        graphemeLines.push(line);
-        line = [];
-        lineWidth = wordWidth;
-        lineJustStarted = true;
-      } else {
-        lineWidth += additionalSpace;
-      }
 
-      if (!lineJustStarted && !splitByGrapheme) {
-        line.push(infix);
-      }
-      line = line.concat(word);
-
-      infixWidth = splitByGrapheme
-        ? 0
-        : this._measureWord([infix], lineIndex, offset);
-      offset++;
-      lineJustStarted = false;
+    if (minWidth + reservedSpace > this.dynamicMinWidth) {
+      this.dynamicMinWidth = minWidth - additionalSpace + reservedSpace;
     }
 
-    i && graphemeLines.push(line);
-
-    if (largestWordWidth + reservedSpace > this.dynamicMinWidth) {
-      this.dynamicMinWidth = largestWordWidth - additionalSpace + reservedSpace;
-    }
-    return graphemeLines;
+    // layout
+    return data.reduce(
+      ({ lines, lineWidth }, { graphemes, infix, infixWidth, width }, i) => {
+        // `i === 0` => `infixWidth === 0`
+        const lineWidthAfter = lineWidth + infixWidth + width;
+        if (i === 0 || lineWidthAfter - additionalSpace > maxWidth) {
+          // push a new line, we spread to protect `data` from mutation
+          lines.push([...graphemes]);
+          lineWidth = width;
```

Comment:
```suggestion
          lineWidth = infixWidth + width;
```

If someone decides to override `measureLineForWrapping` and return an infix we should respect it.

something like:
```
some words some words some words some words 
some words some words ....
    a line with an infix at the start
```

Reply:
and then we should push the infix if it exists

Reply:
Yes, let's abandon this idea.

> measureLines shouldn't change the lines imho, if someone does that is going on its own road with no guarantee of support

I don't understand this